lib/src/modules/apply.cc

```diff
@@ -80,43 +90,51 @@ begin
   # directories we configured earlier.
   Puppet.settings.use(:main)
 
-  ## TODO: ENABLE PLUGINS
+  # This is the way bolt does it
   # Tempfile.open('plugins.tar.gz') do |plugins|
   #   File.binwrite(plugins, Base64.decode64(args['plugins']))
   #   user = Etc.getpwuid.nil? ? Etc.getlogin : Etc.getpwuid.name
   #   Puppet::ModuleTool::Tar.instance.unpack(plugins, moduledir, user)
   # end
 
-  env = Puppet.lookup(:environments).get('production')
-  # Needed to ensure features are loaded
-  env.each_plugin_directory do |dir|
-    $LOAD_PATH << dir unless $LOAD_PATH.include?(dir)
+  # FAILED attempt. Being in an isolated environment breaks this approach
+  # Do plugin sync with plugin face
+  #remote_env_for_plugins = Puppet::Node::Environment.remote(args['environment'])
+  #pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
+  #Puppet[:plugindest] = args['plugin_cache']
+  #Puppet[:pluginfactdest] = args['plugin_cache']
+  #Puppet.override(:http_pool => pool) do
+  #  begin
+  #    handler = Puppet::Configurer::PluginHandler.new()
+  #    handler.download_plugins(remote_env_for_plugins)
+  #  ensure
+  #    pool.close
+  #  end
+  #end
+
+  # Queston about caching... It appears that the file resource type wont need to fully re-download every file in the destdirs. So all the CPP code needs to do 
+  # is to make sure a cache dir named after an environment is available and puppet will do the rest? Does this approach make sense? IE cache by env with a TTL.
+  # HACK: todo figure out a better way, specifically:
+  # 1. Is shelling out (spawning a new proccess) under a leatherman scripting container OK and will it in any way contaminate state in the rest of the shim script
+  # 2. The bolt plugin method packages up the plugins a bit differently than plugin_sync (they are effectively extracted from modules), is copying them
```

Comment:
The cache should be the concatenation of `environment`-`version`.

Reply:
Given we do not have `version` plumbed through puppetserver yet we need to just cache on enviroment for now.